ogr/ogrsf_frmts/xodr/ogr_xodr.h

```diff
@@ -0,0 +1,261 @@
+    std::vector<odr::LaneSection> laneSections;
+    std::vector<std::string> laneRoadIDs;
+    std::vector<odr::Mesh3D> laneMeshes;
+
+    std::vector<odr::Line3D> laneLinesInner;
+    std::vector<odr::Line3D> laneLinesOuter;
+
+    std::vector<odr::RoadMark> roadMarks;
+    std::vector<odr::Mesh3D> roadMarkMeshes;
+
+    std::vector<odr::RoadObject> roadObjects;
+    std::vector<odr::Mesh3D> roadObjectMeshes;
+
+    std::vector<odr::RoadSignal> roadSignals;
+    std::vector<odr::Mesh3D> roadSignalMeshes;
+};
+
+/*--------------------------------------------------------------------*/
+/*---------------------  Layer declarations  -------------------------*/
+/*--------------------------------------------------------------------*/
+
+class OGRXODRLayer : public OGRLayer
+{
+  private:
+    virtual void ResetReading() override;
+
+    virtual OGRFeatureDefn *GetLayerDefn() override
+    {
+        return featureDefn;
+    }
+
+    virtual int TestCapability(const char *) override
+    {
+        return FALSE;
+    }
+
+    /**
+     * Initializes XODR road elements and iterators.
+    */
+    void resetRoadElementIterators();
+
+  protected:
+    VSILFILE *file;
+    RoadElements roadElements;
+    OGRSpatialReference spatialRef;
+    bool dissolveTIN;
+    OGRFeatureDefn *featureDefn;
+
+    /* Unique feature ID which is automatically incremented for any new road feature creation. */
+    int nNextFID;
+
+    std::map<std::string, odr::Road>::iterator roadIter;
+    std::vector<odr::Line3D>::iterator referenceLineIter;
+
+    std::vector<odr::Lane>::iterator laneIter;
+    std::vector<odr::LaneSection>::iterator laneSectionIter;
+    std::vector<std::string>::iterator laneRoadIDIter;
+    std::vector<odr::Mesh3D>::iterator laneMeshIter;
+
+    std::vector<odr::Line3D>::iterator laneLinesInnerIter;
+    std::vector<odr::Line3D>::iterator laneLinesOuterIter;
+
+    std::vector<odr::RoadMark>::iterator roadMarkIter;
+    std::vector<odr::Mesh3D>::iterator roadMarkMeshIter;
+
+    std::vector<odr::RoadObject>::iterator roadObjectIter;
+    std::vector<odr::Mesh3D>::iterator roadObjectMeshesIter;
+
+    std::vector<odr::RoadSignal>::iterator roadSignalIter;
+    std::vector<odr::Mesh3D>::iterator roadSignalMeshesIter;
+
+    /**
+     * Completes feature class definition with all specific attributes and geometry type
+     * according to layer type.
+    */
+    virtual void defineFeatureClass() = 0;
+
+    /**
+     * Builds an ordinary TIN from libOpenDRIVE's mesh.
+    */
+    OGRTriangulatedSurface triangulateSurface(odr::Mesh3D mesh);
+
+  public:
+    /**
+     * \param dissolveTriangulatedSurface True if original triangulated surface meshes from 
+     * libOpenDRIVE are to be dissolved into simpler geometries.
+     * Only applicable for layer types derived from meshes.
+    */
+    // TODO For lower memory consumption maybe better pass xodrRoadElements by reference?
+    OGRXODRLayer(RoadElements xodrRoadElements, std::string proj4Defn);
```

Comment:
The coding standard in the code base is to pass objects as much as possible as const&. At least for std::string. Not sure how doable it is for your OpenDrive objects